enhancements/machine-api/control-plane.md

```diff
@@ -0,0 +1,281 @@
+8 - Delete machine. Go to 1. (Race between the node going away and Cluster Etcd Operator re-adding the member?)
+
+#### Declarative Vertical scaling (scale out + scale in)
+1 - Watch changes to the Control Plane provideSpec
+2 - Fetch all existing control plane Machine resources by ownerRef. Adopt any other machine having a targeted label.
+3 - Fetch all machines with old providerSpec.
+4 - If any machine has old providerSpec then signal controller as "needs upgrade".
+5 - If any machine has "replaced" annotation then trigger scale in workflow (starting in 4) and requeue. Else:
+6 - Pick oldest machine in more populated domain failure.
+7 - Trigger scale out workflow (starting in 4). Set "replaced" annotation. Requeue.
+This is effectively a rolling upgrade with maxUnavailable 0 and maxSurge 1.
+
+#### Self healing (MHC + reconciling)
+1 - The Controller should always reconcile by removing Etcd members for voluntary Machine disruptions, i.e machine deletion.
+2 - At creation time, unless indicated otherwise `EnableAutorepair=true` will be default. The controller will create a Machine Health Checking resource with an ownerRef which will monitor the Control Plane Machines. This will request unhealthy machines to be deleted. Related https://github.com/openshift/machine-api-operator/pull/543.
+3 - The controller will ensure the `maxUnhealthy` value in the MHC resource is set a to known integer to prevent farther remediation from happening during scenarios where quorum could be violated. E.g 1 for cluster with 3 members or 2 for a cluster with 5 members.
+
+#### Bootstrapping (scale out)
+Currently during a regular IPI bootstrapping process the installer uses Terraform to create a bootstrapping instance and 3 master instances. Then it creates Machine resources to "adopt" the existing master instances. In the past Etcd quorum needed to be reached between the three of them before having storage available for the control plane to run self hosted and so for the CVO to run its payload.
+
+The Cluster Etcd Operator introduced support for a single member Etcd cluster to be available quickly on the bootstrapping machine. This lets the CVO to be deployed much faster while new Etcd members are added organically as masters come up.
+
+This proposes dropping terraform for creating masters instances in favour of letting the installer to define a Control Plane resource that scales from zero to 3 replicas as soon as the CVO runs the Machine API Operator. Alternatively if this happen to not be doable because of chicken-egg issues, we could keep the current workflow and include an additional step to create the Control Plane resource which would just adopt existing Master machines.
+
+#### API
+```
+type ControlPlane struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   ControlPlaneSpec   `json:"spec,omitempty"`
+	Status ControlPlaneStatus `json:"status,omitempty"`
+}
+
+type ControlPlaneSpec struct {
```

Comment:
Should this use a `MachineTemplateSpec` in the same way a `MachineSet` does so that they look similar + get objectmeta label copying?

Reply:
why would we expose anything else than we really need at the API? Couldn't anything other than the providerSpec be generated by the controller?

Reply:
My thought was that users may want to add labels to their control-plane machines, that's all we use the objectmeta for on the the `MachineTemplateSpec` within `MachineSet`s right?